service/tablet_allocator.cc

```diff
@@ -758,10 +759,19 @@ class load_balancer {
 
         // Prepare plans for each DC separately and combine them to be executed in parallel.
         for (auto&& dc : topo.get_datacenters()) {
-            auto dc_plan = co_await make_plan(dc);
-            auto level = dc_plan.size() > 0 ? seastar::log_level::info : seastar::log_level::debug;
-            lblogger.log(level, "Prepared {} migrations in DC {}", dc_plan.size(), dc);
-            plan.merge(std::move(dc_plan));
+            if (_db.get_config().rf_rack_valid_keyspaces()) {
```

Comment:
Is it good to trust the value of a configuration option? Can effective_replication_map maintain it instead?

Reply:
Does the configuration option being set mean that the tablet distribution actually follows it?

Do we have a test that enabling it on non-compliant distributions fails to boot?

Also, we need to set it to true in scylla.yaml so it's enforced in new clusters.

Reply:
I don't understand. What happens if there are non-conforming tables, but the configuration says we are conforming?

Reply:
> I don't understand. What happens if there are non-conforming tables, but the configuration says we are conforming?

Tablet replicas will stick to their racks. Which was the intent behind rf-rack-valid.

If tables don't conform, there can be side effects. Merge may not be able to proceed if #racks > RF. Or imbalance may form if some racks had more replicas of a given table than others.

It can be recovered, by disabling the config option and adjusting topology back to conformity.

There is a problem with RF=1, which is considered rack-valid even if RF<#racks. Before we have rack-list based RF, we probably should still balance such keyspaces across racks. There are no replication constraints due to racks for them, so they should not be prone to the performance problems.